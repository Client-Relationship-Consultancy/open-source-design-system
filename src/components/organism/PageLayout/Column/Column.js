import React from "react"
import styled from "styled-components"
import PropTypes from "prop-types"
import { media } from "mediaQueries"

const StyledColumn = styled.div`
  width: 100%;
  grid-column: ${props => {
    const { span = ["auto", 12] } = props
    return `${span[0]} / span ${span[1]}`
  }};
  grid-row: ${props => {
    const { row = ["auto", 1] } = props
    // Check if number or string
    // If number, prefix it with span, else just return string (for use case of end-line)
    // check css-grid for documentation
    const endingValue = typeof row[1] === "number" ? `span ${row[1]}` : row[1]
    return `${row[0]} / ${endingValue}`
  }};
  ${media.tabletPortrait`
    grid-column: ${props => {
      const { spanTabletPortrait } = props
      return spanTabletPortrait ? `${spanTabletPortrait[0]} / span ${spanTabletPortrait[1]}` : null
    }};
  `}
  ${media.tabletLandscape`
    grid-column: ${props => {
      const { spanTabletLandscape } = props
      return spanTabletLandscape
        ? `${spanTabletLandscape[0]} / span ${spanTabletLandscape[1]}`
        : null
    }};
  `}
  ${media.desktop`
    grid-column: ${props => {
      const { spanDesktop } = props
      return spanDesktop ? `${spanDesktop[0]} / span ${spanDesktop[1]}` : null
    }};
  `}
`
StyledColumn.displayName = "StyledColumn"

const Column = props => {
  const {
    children,
    span,
    spanTabletPortrait,
    spanTabletLandscape,
    spanDesktop,
    row,
    className,
    id,
  } = props
  return (
    <StyledColumn
      id={id}
      span={span}
      spanTabletPortrait={spanTabletPortrait}
      spanTabletLandscape={spanTabletLandscape}
      spanDesktop={spanDesktop}
      row={row}
      className={className}
    >
      {children}
    </StyledColumn>
  )
}
Column.displayName = "Column"
Column.propTypes = {
  children: PropTypes.node,
  span: PropTypes.array,
  spanTabletPortrait: PropTypes.array,
  spanTabletLandscape: PropTypes.array,
  spanDesktop: PropTypes.array,
  row: PropTypes.array,
  className: PropTypes.string,
  id: PropTypes.string,
}

export default Column
