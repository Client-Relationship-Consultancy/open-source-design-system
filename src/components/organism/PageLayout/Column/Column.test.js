import React from "react"
import { shallow, mount } from "enzyme"
import "jest-styled-components"
import toJson from "enzyme-to-json"
import Column from "./Column"

describe("Column component test", () => {
  it("should match snapshot", () => {
    const component = shallow(<Column>snapshot</Column>)
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })

  it("should render children correctly", () => {
    const children = "this is the child"
    const component = shallow(<Column>{children}</Column>)
    expect(component.text()).toEqual(children)
  })

  it("should check default props", () => {
    const component = mount(<Column />)
    expect(component).toHaveStyleRule("grid-column", "auto / span 12")
    expect(component).toHaveStyleRule("grid-row", "auto / span 1")
  })

  it("should be able to pass string or number to row props, number should prefix it with span", () => {
    const component = mount(<Column row={[1, 3]} />)
    expect(component).toHaveStyleRule("grid-row", "1 / span 3")
    component.setProps({ row: [1, "end-line"] })
    expect(component).toHaveStyleRule("grid-row", "1 / end-line")
  })

  it("should receive tabletPortrait media queries", () => {
    const component = mount(<Column spanTabletPortrait={[1, 3]} />)
    expect(component).toHaveStyleRule("grid-column", "1 / span 3", {
      media: "all and (min-width: 48em)",
    })
  })

  it("should receive tabletLandscape media queries", () => {
    const component = mount(<Column spanTabletLandscape={[1, 3]} />)
    expect(component).toHaveStyleRule("grid-column", "1 / span 3", {
      media: "all and (min-width: 60em)",
    })
  })

  it("should receive desktop media queries", () => {
    const component = mount(<Column spanDesktop={[1, 3]} />)
    expect(component).toHaveStyleRule("grid-column", "1 / span 3", {
      media: "all and (min-width: 75em)",
    })
  })
})
