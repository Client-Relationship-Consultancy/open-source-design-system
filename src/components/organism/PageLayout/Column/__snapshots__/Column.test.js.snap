// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column component test should match snapshot 1`] = `
<StyledColumn>
  snapshot
</StyledColumn>
`;
